File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res, subset = [], []
        def dfs(o, c):
            if o == c and c == n:
                res.append("".join(subset))
            if o < n:
                subset.append("(") 
                dfs(o + 1, c)
                subset.pop()
            if c < o:
                subset.append(")")
                dfs(o, c + 1)
                subset.pop()
        dfs(0, 0)
        return res

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        Stack<Character> s = new Stack<>();
        List<String> res = new ArrayList<>();
        backtrack(0, 0, n, s, res);
        return res;
    }
    public void backtrack(
        int o,
        int c,
        int n,
        Stack<Character> s,
        List<String> res
    ){
        if(o == c && c == n){
            Iterator it = s.iterator();
            StringBuffer bf = new StringBuffer();
            while(it.hasNext()){
                bf.append(it.next());
            }
            res.add(bf.toString());
        }
        if(o < n){
            s.add('(');
            backtrack(o + 1, c, n, s, res);
            s.pop();
        }
        if(c < o){
            s.add(')');
            backtrack(o, c + 1, n, s, res);
            s.pop();
        }
    }
}